lib/services/local/trust.go

```diff
@@ -482,16 +730,44 @@ func (s *CA) GetTrustedClusters(ctx context.Context) ([]types.TrustedCluster, er
 
 // DeleteTrustedCluster removes a TrustedCluster from the backend by name.
 func (s *CA) DeleteTrustedCluster(ctx context.Context, name string) error {
+	return s.DeleteTrustedClusterInternal(ctx, name, nil /* no cert authorities */)
+}
+
+// DeleteTrustedClusterInternal removes a trusted cluster and associated resources atomically.
+func (s *CA) DeleteTrustedClusterInternal(ctx context.Context, name string, caIDs []types.CertAuthID) error {
 	if name == "" {
 		return trace.BadParameter("missing trusted cluster name")
 	}
-	err := s.Delete(ctx, backend.NewKey(trustedClustersPrefix, name))
-	if err != nil {
-		if trace.IsNotFound(err) {
+
+	for _, id := range caIDs {
+		if err := id.Check(); err != nil {
+			return trace.Wrap(err)
+		}
+
+		if id.DomainName != name {
+			return trace.BadParameter("ca %q does not belong to trusted cluster %q", id.DomainName, name)
+		}
+	}
+
+	condacts := []backend.ConditionalAction{
+		{
+			Key:       backend.NewKey(trustedClustersPrefix, name),
+			Condition: backend.Exists(),
+			Action:    backend.Delete(),
+		},
+	}
+
+	condacts = append(condacts, s.deleteCertAuthoritiesCondActs(caIDs)...)
```

Comment:
Aren't we at a slight risk of leaving a CA behind if it was created after we fetch the currently existing ones? Or is that under the umbrella of "manual cert authority manipulation is liable to result in weird states so admin discretion is advised"?

Auth version mismatches notwithstanding (CAs are stored by type and name so any unsupported CA effectively doesn't exist, unfortunately), we could've just deleted all known CA types with the given name unconditionally, right?

Reply:
`auth/trustedcluster.go` passes in IDs for all known CA types, and all CA deletes are unconditional.

Reply:
I misread, sorry.